ocaml/database/db_cache_types.ml

```diff
@@ -350,10 +350,10 @@ let remove_from_set key t =
   Schema.Value.Set (List.filter (fun x -> x <> key) t)
 
 exception Duplicate
-let add_to_map key value t =
+let add_to_map idempotent key value t =
   let t = Schema.Value.Unsafe_cast.pairs t in
-  if List.mem key (List.map fst t) then raise Duplicate;
-  Schema.Value.Pairs ((key, value) :: t)
+  if List.mem_assoc key t && (not idempotent || List.assoc key t <> value) then raise Duplicate;
```

Comment:
I don't think it's quite so clear cut here - checking the `idempotent` condition is only relevant if the `mem_assoc` test is true, and that's expected to be quite rare. Also, in the default case here `idempotent` is false, hence we don't do the List.assoc.